File: 11-container-with-most-water/container-with-most-water.rs
/*
two pointers:
left 0 
right n-1

theres a space of 1 unit between each line 

area = min(left, right) * max(left, right)-min(left, right)
area = max(area,max_area)
if left < right 
    go next left
else 
    go back right

height >=0
if height == 0
*/

use std::cmp::{min, max};
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let n: i32 = height.len() as i32;
        let mut li = 0;
        let mut ri = n-1;
        let mut max_area = 0;
        while li < ri {
            let left = height[li as usize];
            let right = height[ri as usize];
            let current_area = min(left, right) * (ri-li);
            println!("{:?}, {:?}, {:?}", current_area, li, ri);
            max_area = max(current_area, max_area);
            if left > right{ 
                ri-=1;
            }
            else {
                li+=1;
            }

        }
        return max_area
    }
}

File: 11-container-with-most-water/container-with-most-water.py
"""
ok so the space between each of the lines is 1 basically

this is a maximization problem, we can think of dp, greedy, 

i think this can be solved with dp.... no because ach one doesnt build up to the overall solution, i think this cannot be divided
into subproblems, like the solution changes each time we add smth

i think that remembering from this solution the intuition was that the maximum area is either left to right or right to left

ok so the maximum is from the longest to the longest and fartest from it 

ok so obviously the first thing is brute force, calculating all the posibilities but i **know** theres something more efficient

ok the thing is we can get the biggest left to right, and then right to left
"""
class Solution:
    def maxArea(self, height: List[int]) -> int:
        n = len(height)
        left = 0
        right =  n - 1
        max_area = 0
        while left < right:
            area = min(height[left], height[right]) * (right - left)
            max_area = max(area, max_area)
            if height[left] > height[right]:
                right -=1
            else:
                left +=1

        return max_area